### PR TITLE
boards: shields: add suport for weact_ov2640_cam_module

### PR DESCRIPTION
**Pull request by CharlesDias**
Add support for OV2640 camera sensor to use with WeAct Studio MiniSTM32H7xx boards.

PR dependency:
- https://github.com/zephyrproject-rtos/zephyr/pull/71463

### CODE REVIEW

**Comment by CharlesDias** [comment]
> I think we should define a binding for the connector (similar to `dsi_lcd_qsh_030` display connector for example) since it's apparently used by many vendors ([#71463 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/71463#discussion_r1713617227)) and use that binding to specify the pins in ov2640 node, in a board-independent way.
> 
> Also, the I2C, DCMI (except `port` child node, and board-independent properties like `sensor`), and DMA nodes should be moved to a board-specific overlay (#77854 does the same for display shield) & then be referenced by generic aliases (like "cam_interface" for DCMI so the sensor can be used with a board that has a cam interface other than DCMI).
> 
> BTW, should the shield be renamed to "ov2640_cam_module" for example?

Thanks for your review, @JarmouniA! 

Since I'm not familiar with the connector binding approach, I'll take a look at it carefully.

**Comment by CharlesDias** [comment]
Hi, @JarmouniA.

> I think we should define a binding for the connector (similar to `dsi_lcd_qsh_030` display connector for example) since it's apparently used by many vendors ([#71463 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/71463#discussion_r1713617227)) and use that binding to specify the pins in ov2640 node, in a board-independent way.

I think it is done! Let me know if I got right your suggestion. 

> BTW, should the shield be renamed to "ov2640_cam_module" for example?

About the shield name, could it be "weact_ov2640_cam_module"? To make it less generic.

**Comment by CharlesDias** [comment]
> My point was that the connector should't be considered a WeAct one, it's used on a lot of boards from other vendors as well (like ESP32-CAM https://components101.com/sites/default/files/inline-images/ESP32-CAM-Schematics.png), so if we are going to make a compatible for it, it should not have WeAct in its name. Same thing for the camera module (https://components101.com/sites/default/files/components/ESP32-CAM-Camera-Module.jpg).

Good observation, @JarmouniA. 

There is one point here to consider: even though the connector is similar, the position of the pins is different from the WeAct schematic https://github.com/WeActStudio/MiniSTM32H7xx/blob/master/HDK/STM32H7xx%20SchDoc%20V11.pdf. I believe this is a problem when creating a generic OV2640 shield version. What you think?

**Comment by CharlesDias** [comment]
I placed the connectors next to each other and noticed that the pins are upside down. Additionally, the ESP32-CAM doesn't connect the PWDN pin, while the WeAct doesn't connect the RESET pin (which is linked to the reset system).

image

**Comment by CharlesDias** [comment]
> > the position of the pins is different from the WeAct schematic
> 
> Ah, good catch, didn't pay attention to this.
> 
> In this case, would it make sense to have a generic compatible for the connector, then specific ones for each vendor, depending on the pins mapping, that include the common one? Or would it be simpler to just keep the current implementation? The same for the cam module.

I think we would keep it simpler as in the current implementation. Is it ok for you?

**Comment by CharlesDias** [comment]
Thank you @JarmouniA for your valuable review!

**Comment by CharlesDias** [comment]
Change the shield name to `weact_ov2640_cam_module`.

**Comment by josuah** [comment]
This introduces a new naming for shields.
As noted here, there seems to be a `zephyr_` prefix for all shields, which might allow to make a difference between names specific to a SoC/Boad/project and those standardized by Zephyr, facilitating review of all standards:
- https://github.com/zephyrproject-rtos/zephyr/pull/77854#discussion_r1740830325

There is a compromise to take in that:

- `dvp` is relatively widespread but roots back to OmniVision
- `dcmi` is a peripheral by ST
- `camera_interface` sounds like the [right name](https://en.wikipedia.org/wiki/Camera_interface) but could be mistaken with MIPI-CSI which also "a camera interface".
- `camif` does not seems as broadly as the vendor terms.

If we try to copy-cat `zephyr_mipi_dsi`, this could be:

- `zephyr_camif` or `zephyr_dvp` or `zephyr_camera_dvp` or zephyr_camera_parallel`
- `zephyr_camera_i2c` or `zephyr_camif_i2c` or `zephyr_dvp_i2c` or `zephyr_camera_parallel_i2c` or `zephyr_mipi_csi_i2c`

Which one sounds the best?

**Comment by CharlesDias** [comment]
> Which one sounds the best?

Hi, @josuah. The `zephyr_camera_dvp` and `zephyr_camera_i2c`sounds great for me!
Thanks for your suggestion!

**Comment by CharlesDias** [comment]
- Change `camera_interface` to `zephyr_camera_dvp`.
- Change `camera_connector_i2c` to `zephyr_camera_i2c`.

**Review by josuah — DISMISSED** [review]
The `remote-endpoint` is present but not actively used yet as not supported by Zephyr devicetree syntax. This should not be any problem to be sitting there AFAIU from #57708.

@JarmouniA I tried to add your other account but seems inactive, it is still possible to make reviews even without collaborator access if you wish.

**Comment by josuah** [comment]
CI complains about `samples/drivers/video/capture_to_lvgl/` not existing, which will be introduced by #71463.

**Review by JarmouniA — APPROVED** [review]

**Review by erwango — DISMISSED** [review]

**Comment by erwango** [comment]
@CharlesDias There are some CI issues to be fixed

**Comment by CharlesDias** [comment]
Hi @erwango. 
This PR https://github.com/zephyrproject-rtos/zephyr/pull/71463 has to be merged first. Could you review it? Thank you!

**Comment by CharlesDias** [comment]
Rebase.

**Comment by CharlesDias** [comment]
Add new MCO configuration as suggested on https://github.com/zephyrproject-rtos/zephyr/pull/71463#discussion_r1773212375

**Comment by CharlesDias** [comment]
Rebase.

File: boards/shields/weact_ov2640_cam_module/Kconfig.shield [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2024 Charles Dias
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_WEACT_OV2640_CAM_MODULE
+	def_bool $(shields_list_contains,weact_ov2640_cam_module)
```
Comment (josuah) [inline boards/shields/weact_ov2640_cam_module/Kconfig.shield:5]: I did not know Zephyr had custom [Kconfig](https://docs.zephyrproject.org/latest/build/kconfig/preprocessor-functions.html) macros!

File: boards/shields/weact_ov2640_cam_module/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+   :align: center
+   :alt: OV2640 Camera Sensor
+
+More information about the OV2640 camera sensor can be found on the
+`MiniSTM32H7xx GitHub`_ and in the `OV2640 datasheet`_.
+
+Requirements
+************
+
+Your board needs to have a ``dcmi`` device tree label to work with this shield.
```
Comment (josuah) [inline boards/shields/weact_ov2640_cam_module/doc/index.rst:24]: Is it now `zephyr_camera_dvp`?
https://github.com/zephyrproject-rtos/zephyr/compare/d73a2d869c5c426f64aea6860b63205a42a1fbbf..76759c81c076ea47aebe244b2f5daa940ae32eca#diff-a04707eebe85f0f2eef1831b4061cffe823f57bbf4a2ac8f549d08a167f3e106L28

Maybe it was to mean that it is only expected to work with DCMI in the current state... in which case "devicetree label".

Also, `device tree` -> `devicetree` according to this search, also also used interchangeably it seems...
https://docs.zephyrproject.org/latest/gsearch.html?q=device+tree&check_keywords=yes&area=default#gsc.tab=0&gsc.q=device%20tree&gsc.page=1

**Review by josuah — CHANGES_REQUESTED** [review]
Some very minor documentation edits spotted but everything else looks ready to me!

File: boards/shields/weact_ov2640_cam_module/Kconfig.shield (hunk 1 above)
Comment (CharlesDias) [reply boards/shields/weact_ov2640_cam_module/Kconfig.shield:5]: Zephyr is full of surprises! :)

**Review by JarmouniA — APPROVED** [review]

**Comment by CharlesDias** [comment]
Update the [samples/drivers/video/capture_to_lvgl/sample.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/bda49ac20bcd88c2381ecdc19894630d5b7d4649/samples/drivers/video/capture_to_lvgl/sample.yaml) file to include this shield.

**Comment by CharlesDias** [comment]
Hello, everyone.

I am trying to understand why the CI is failing! Do you have any suggestions for me? 

Another point: should I update the [samples/drivers/video/capture_to_lvgl/sample.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/bda49ac20bcd88c2381ecdc19894630d5b7d4649/samples/drivers/video/capture_to_lvgl/sample.yaml) to include this shield? If that's the case, should it be like this?

```
sample:
  name: Video capture to LVGL
tests:
  sample.video.capture_to_lvgl:
    tags:
      - video
      - shield
      - samples
    extra_args:
      - platform:mini_stm32h743:SHIELD=weact_ov2640_cam_module
    harness: console
    harness_config:
      fixture: fixture_camera
      type: multi_line
      ordered: true
      regex:
        - "Device name"
        - "Format"
        - "Capture started"
    platform_allow:
      - mini_stm32h743
    depends_on: video
    integration_platforms:
      - mini_stm32h743
```

**Comment by CharlesDias** [comment]
> @CharlesDias
> 
> ```
> extra_args:
>       - platform:mini_stm32h743:SHIELD=weact_ov2640_cam_module
> ```
> 
> this does not look right.
> 
> See
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/3a15e37b5a9a0d17c111eecb3a1803c2ab20cd5d/samples/drivers/display/sample.yaml#L136

Thanks, @JarmouniA! Do I have to change also from `sample.video.capture_to_lvgl:` to `sample.video.capture_to_lvgl.shield.weact_ov2640_cam_module:`?

**Comment by CharlesDias** [comment]
I used as a model the https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/video/capture/sample.yaml file.

**Comment by CharlesDias** [comment]
Not yet, @JarmouniA! There is something more! Twister is mysterious to me! :)

File: boards/shields/weact_ov2640_cam_module/Kconfig.defconfig [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2024 Charles Dias
+# SPDX-License-Identifier: Apache-2.0
+
+if SHIELD_WEACT_OV2640_CAM_MODULE
+
+config VIDEO_HFLIP
+	default n
+
+config VIDEO_VFLIP
+	default n
```
Comment (josuah) [inline boards/shields/weact_ov2640_cam_module/Kconfig.defconfig:10]: I think this is where Twister complains now: it needs something like `bool Description of this`.

Comment (josuah) [reply boards/shields/weact_ov2640_cam_module/Kconfig.defconfig]:
```
warning: VIDEO_HFLIP (defined at boards/shields/weact_ov2640_cam_module/Kconfig.defconfig:6) defined without a type

warning: VIDEO_VFLIP (defined at boards/shields/weact_ov2640_cam_module/Kconfig.defconfig:9) defined without a type

error: Aborting due to Kconfig warnings
```

**Comment by CharlesDias** [comment]
Removed the [boards/shields/weact_ov2640_cam_module/Kconfig.defconfig](https://github.com/zephyrproject-rtos/zephyr/compare/5cf0686b9c4423d8e672033fb934cca6c18bac6d..54760b90635dc46f16a250fc298fe638fc2f5776#diff-ea21e0c2c49182c2419f4292fa61b0efb8bf48cd8593319fda544d69736d800d) file.

**Comment by CharlesDias** [comment]
Rebase

File: boards/shields/weact_ov2640_cam_module/weact_ov2640_cam_module.overlay [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	status = "okay";
+	sensor = <&ov2640>;
+	bus-width = <8>;
+	hsync-active = <0>;
+	vsync-active = <0>;
+	pixelclk-active = <1>;
+	capture-rate = <1>;
+
+	port {
+		dcmi_ep_in: endpoint {
```
Comment (josuah) [inline boards/shields/weact_ov2640_cam_module/weact_ov2640_cam_module.overlay:40]: Maybe this would need an alias to be as generic as the `zephyr_camera_dvp`? `zephyr_camera_dvp_ep_in`? Or less verbose `zephyr_camera_dvp_in`? Or something else?

File: boards/shields/weact_ov2640_cam_module/weact_ov2640_cam_module.overlay [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	clock-frequency = <I2C_BITRATE_FAST>;
+
+	ov2640: ov2640@30 {
+		compatible = "ovti,ov2640";
+		reg = <0x30>;
+		status = "okay";
+
+		port {
+			ov2640_ep_out: endpoint {
+				remote-endpoint = <&dcmi_ep_in>;
```
Comment (josuah) [inline boards/shields/weact_ov2640_cam_module/weact_ov2640_cam_module.overlay:24]: See comment below about what to pick to make it generic.

**Review by josuah — DISMISSED** [review]
It looks in good condition to be merged.
The last little details that might have been missed will be a lot easier to fix once an alternative camera module is to be supported for this board, or once an alternative board tries to use that camera module.

Thank you for this integration!

**Review by JarmouniA — APPROVED** [review]

**Comment by CharlesDias** [comment]
Hi, everyone.

What you think about the use of `CONFIG_VIDEO_HFLIP=y` and `CONFIG_VIDEO_VFLIP=y`symbols? In this way, the shield depends on these symbols that are relevant only for sample/drivers/video/capture_to_lvgl code. Does this make sense? Is there another way? Thanks!

**Comment by CharlesDias** [comment]
Change port label from `dcmi_ep_in` to `zephyr_camera_dvp_in`.

**Review by josuah — DISMISSED** [review]

**Review by JarmouniA — APPROVED** [review]

**Comment by josuah** [comment]
> What you think about the use of `CONFIG_VIDEO_HFLIP=y` and `CONFIG_VIDEO_VFLIP=y` symbols inside of [boards/shields/weact_ov2640_cam_module/boards/mini_stm32h743.conf](https://github.com/zephyrproject-rtos/zephyr/pull/77949/files#diff-8e3a95d81dbc55b3f05d19e162b50a7488bd57e9ffb66b9e97d214400118e05a)? In this way, the shield depends on these symbols that are relevant only for sample/drivers/video/capture_to_lvgl code. Does this make sense? Is there another way? Thanks!

The FPC can eventually be folded to rotate the camera by 180°, but then that can still be overridden by the user.
This reminds me of [`LIBV4LCONTROL_FLAGS`](https://github.com/philips/libv4l/blob/master/libv4lconvert/control/libv4lcontrol.c#L369-L371) that permits to [rotate a misoriented camera](https://techieboycdo.blogspot.com/2010/08/fix-upside-down-or-inverted-webcam-on.html) at the very last moment.

File: boards/shields/weact_ov2640_cam_module/boards/mini_stm32h743.overlay [hunk 6]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) Copyright (c) 2024 Charles Dias
```
Comment (kartben) [inline boards/shields/weact_ov2640_cam_module/boards/mini_stm32h743.overlay:2]: typo

File: boards/shields/weact_ov2640_cam_module/weact_ov2640_cam_module.overlay [hunk 7]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) Copyright (c) 2024 Charles Dias
```
Comment (kartben) [inline boards/shields/weact_ov2640_cam_module/weact_ov2640_cam_module.overlay:2]: typo

**Review by kartben — DISMISSED** [review]

**Review by JarmouniA — APPROVED** [review]

**Review by josuah — DISMISSED** [review]

File: samples/drivers/video/capture_to_lvgl/Kconfig
Comment (kartben) [inline samples/drivers/video/capture_to_lvgl/Kconfig:1]: can you put this in a different commit please?

**Review by kartben — CHANGES_REQUESTED** [review]
one small comment, looks good otherwise!

**Review by kartben — APPROVED** [review]

**Review by josuah — APPROVED** [review]
